protocols/imap/src/main/scala/org/apache/james/imap/decode/parser/SearchCommandParser.scala

```diff
@@ -0,0 +1,683 @@
+    expectCharacters("undeleted", alreadyConsumed)(request)
+    UnDeleted()
+  }
+
+  @throws[DecodingException]
+  private def unflagged(request: ImapRequestLineReader, alreadyConsumed: Int): SearchKey = {
+    expectCharacters("unflagged", alreadyConsumed)(request)
+    UnFlagged()
+  }
+
+  @throws[DecodingException]
+  private def unanswered(request: ImapRequestLineReader, alreadyConsumed: Int): SearchKey = {
+    expectCharacters("unanswered", alreadyConsumed)(request)
+    UnAnswered()
+  }
+
+  @throws[DecodingException]
+  private def younger(request: ImapRequestLineReader, alreadyConsumed: Int): SearchKey = {
+    expectCharacters("younger ", alreadyConsumed)(request)
+    Younger(request.nzNumber)
+  }
+
+  @throws[DecodingException]
+  private def older(request: ImapRequestLineReader, alreadyConsumed: Int): SearchKey = {
+    expectCharacters("older ", alreadyConsumed)(request)
+    Older(request.nzNumber)
+  }
+
+  @throws[DecodingException]
+  private def or(session: ImapSession, request: ImapRequestLineReader, charset: Charset, alreadyConsumed: Int): SearchKey = {
+    expectCharacters("or ", alreadyConsumed)(request)
+    val firstKey = searchKey(session, request, charset, isFirstToken = false)
+    nextIsSpace(request)
+    val secondKey = searchKey(session, request, charset, isFirstToken = false)
+    Or(firstKey, secondKey)
+  }
+
+  @throws[DecodingException]
+  private def not(session: ImapSession, request: ImapRequestLineReader, charset: Charset, alreadyConsumed: Int): SearchKey = {
+    expectCharacters("not ", alreadyConsumed)(request)
+    val nextKey = searchKey(session, request, charset, isFirstToken = false)
+    Not(nextKey)
+  }
+
+  @throws[DecodingException]
+  private def newOperator(request: ImapRequestLineReader, alreadyConsumed: Int): SearchKey = {
+    expectCharacters("new", alreadyConsumed)(request)
+    New()
+  }
+
+  @throws[DecodingException]
+  private def recent(request: ImapRequestLineReader, alreadyConsumed: Int): SearchKey = {
+    expectCharacters("recent", alreadyConsumed)(request)
+    Recent()
+  }
+
+  @throws[DecodingException]
+  private def seen(request: ImapRequestLineReader, alreadyConsumed: Int): SearchKey = {
+    expectCharacters("seen", alreadyConsumed)(request)
+    Seen()
+  }
+
+  @throws[DecodingException]
+  private def draft(request: ImapRequestLineReader, alreadyConsumed: Int): SearchKey = {
+    expectCharacters("draft", alreadyConsumed)(request)
+    Draft()
+  }
+
+  @throws[DecodingException]
+  private def deleted(request: ImapRequestLineReader, alreadyConsumed: Int): SearchKey = {
+    expectCharacters("deleted", alreadyConsumed)(request)
+    Deleted()
+  }
+
+  @throws[DecodingException]
+  private def b(request: ImapRequestLineReader, charset: Charset, alreadyConsumed: Int): SearchKey =
+    consumeAndCap(request) match {
+      case 'C' =>
+        bcc(request, charset, alreadyConsumed + 1)
```

Comment:
What this (technical) +1 means?

Reply:
every time I consumer a char (with `consume*` methods, I increment the `alreadyConsumer` counter to allow `expectCharacters` to match only characters that didn't matched yet

Reply:
Can we find a better way of expressing this in the code?

Reply:
is it better than before? If not, I can revert that part.
If yes, please way the refactoring introducing a proper grammar parser (:

Reply:
It is better than before.